app/ante/reject_msgs.go

```diff
@@ -70,51 +86,40 @@ func (rmd RejectMessagesDecorator) checkMsg(ctx sdk.Context, msg sdk.Msg, nested
 		)
 	}
 
-	if nestedMsgs >= maxNestedMsgs {
-		return fmt.Errorf("found more nested msgs than permitted. Limit is : %d", maxNestedMsgs)
+	typeURL := sdk.MsgTypeURL(msg)
+	for _, pred := range rmd.predicates {
+		if pred(typeURL, depth) {
+			return gerrc.ErrInvalidArgument.Wrapf("disabled: %s", typeURL)
+		}
 	}
 
-	innerMsgs, err := extractMsgs(msg)
-	if err != nil {
-		return err
-	}
-	switch concreteMsg := msg.(type) {
+	var err error
+	var inner []sdk.Msg
+
+	switch m := msg.(type) {
 	case *authz.MsgExec:
-		nestedMsgs++
-		if err := rmd.checkMsgs(ctx, innerMsgs, nestedMsgs); err != nil {
-			return err
-		}
+		inner, err = m.GetMessages()
+	case *govtypesv1.MsgSubmitProposal:
+		inner, err = m.GetMsgs()
+	case *group.MsgSubmitProposal:
+		inner, err = m.GetMsgs()
+	default:
 	case *authz.MsgGrant:
-		authorization, err := concreteMsg.GetAuthorization()
+		authorization, err := m.GetAuthorization()
 		if err != nil {
 			return err
 		}
-		url := authorization.MsgTypeURL()
-		if _, ok := rmd.disabledMsgTypeURLs[url]; ok {
-			return fmt.Errorf("granting disabled msg type: %s is not allowed", url)
-		}
-	case *govtypesv1.MsgSubmitProposal:
-		nestedMsgs++
-		if err := rmd.checkMsgs(ctx, innerMsgs, nestedMsgs); err != nil {
-			return err
-		}
-	case *group.MsgSubmitProposal:
-		nestedMsgs++
-		if err := rmd.checkMsgs(ctx, innerMsgs, nestedMsgs); err != nil {
-			return err
+		typeURL = authorization.MsgTypeURL()
+		for _, pred := range rmd.predicates {
+			if pred(typeURL, depth) {
+				return gerrc.ErrInvalidArgument.Wrapf("disabled grant: %s", typeURL)
+			}
 		}
-	default:
 	}
 
-	return nil
-}
-
-func extractMsgs(msg any) ([]sdk.Msg, error) {
-	if msgWithMsgs, ok := msg.(interface{ GetMsgs() ([]sdk.Msg, error) }); ok {
-		return msgWithMsgs.GetMsgs()
-	}
-	if msgWithMessages, ok := msg.(interface{ GetMessages() ([]sdk.Msg, error) }); ok {
-		return msgWithMessages.GetMessages()
+	if err != nil {
+		return err
 	}
-	return nil, nil
+
+	return rmd.checkMsgs(ctx, inner, depth+1)
```

Comment:
not sure what the stop cause for this recursion. if no inner found
probably `return nil` on the `default:` case?

Reply:
https://github.com/dymensionxyz/dymension/blob/2ca693f194a422df8a2d75859512fb2b7c5e4de6/app/ante/reject_msgs.go#L73